File: Stack/python/stack.py
import typing

T = typing.TypeVar("T")


class Node:

    def __init__(
            self, element: T,
            next_: typing.Optional["Node"] = None,
            previous: typing.Optional["Node"] = None
    ) -> None:
        self._element = element
        self._next = next_
        self._previous = previous

    @property
    def next(self) -> typing.Union["Node", None]:
        return self._next

    @next.setter
    def next(self, node: "Node") -> None:
        self._next = node
        if node is not None:
            node._previous = self

    @property
    def previous(self) -> typing.Union["Node", None]:
        return self._previous

    @previous.setter
    def previous(self, node: "Node") -> None:
        self._previous = node
        if node is not None:
            node._next = self

    @property
    def element(self) -> T:
        return self._element

    @element.setter
    def element(self, element: T) -> None:
        self._element = element

    def __repr__(self) -> str:
        return (f"{self.__class__.__name__}(element={self._element}, "
                f"next={self._next._element if self._next else None}, "
                f"previous={self._previous._element if self._previous else None})")


class DoublyLinkedList:

    def __init__(
            self,
            head: typing.Optional[Node] = None,
            tail: typing.Optional[Node] = None
    ) -> None:
        self._head = head
        self._tail = tail
        self._size = 0
        self._next: typing.Union[Node, None] = None

    @property
    def size(self) -> int:
        return self._size

    @property
    def head(self) -> typing.Union[Node, None]:
        return self._head

    @head.setter
    def head(self, node: Node) -> None:
        if self._head is not None:
            self._head.previous = node
        self._head = node

    @property
    def tail(self) -> typing.Union[Node, None]:
        return self._tail

    @tail.setter
    def tail(self, node: Node) -> None:
        if self._tail is not None:
            self._tail.next = node
        self._tail = node

    @property
    def empty(self) -> bool:
        return self._size < 1

    def append(self, node: Node) -> None:
        self.tail = node
        if self._head is None:
            self.head = node
        self._size += 1

    def prepend(self, node: Node) -> None:
        self.head = node
        if self._tail is None:
            self.tail = node
        self._size += 1

    def remove(self, node: Node) -> None:
        next_ = node.next
        previous = node.previous
        if next_ is not None:
            next_.previous = previous
        elif previous is not None:
            previous.next = None
        node.next = None
        node.previous = None
        self._size -= 1

    def __iter__(self) -> "DoublyLinkedList":
        self._next = self._head
        return self

    def __next__(self) -> Node:
        if self._next is None:
            raise StopIteration
        n = self._next
        self._next = n.next
        return n


class Stack:

    def __init__(self) -> None:
        self._list = DoublyLinkedList()
        self._iterator: typing.Union[DoublyLinkedList, None] = None

    @property
    def size(self) -> int:
        return self._list.size

    @property
    def empty(self) -> bool:
        return self._list.empty

    def pop(self) -> T:
        head = self._list.head
        if head is None:
            return None
        next_ = head.next
        self._list.remove(head)
        self._list.head = None
        self._list.head = next_
        return head.element

    def top(self) -> T:
        head = self._list.head
        return head.element if head is not None else None

    def push(self, element: T) -> None:
        self._list.prepend(Node(element))

    def __iter__(self) -> "Stack":
        self._iterator = iter(self._list)
        return self

    def __next__(self) -> T:
        return next(self._iterator).element


if __name__ == "__main__":
    stack = Stack()
    stack.push(1)
    stack.push(2)
    stack.push(3)
    print(stack.size)
    print(stack.pop())
    print(stack.size)
    print()
    for e in stack:
        print(e)

File: Stack/c/stack.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

#define NULL_INT -1

typedef struct _Node {
    void *element;
    struct _Node *next;
    struct _Node *previous;
} Node;

const size_t NODE_SIZE = sizeof(Node);

void Node_set_next(Node *parent, Node *child) {
    parent->next = child;
    if (child != NULL) {
        child->previous = parent;
    }
}

void Node_set_previous(Node *parent, Node *child) {
    parent->previous = child;
    if (child != NULL) {
        child->next = parent;
    }
}

void Node_set_element(Node *node, void *element) {
    node->element = element;
}

void Node_print(Node *node) {
    int element = *((int *)node->element);
    int next_element = (node->next != NULL) ? *((int *)node->next->element) : NULL_INT;
    int previous_element = (node->previous != NULL) ? *((int *)node->previous->element) : NULL_INT;
    printf("Node(element=%d, next=%d, previous=%d)\n", element, next_element, previous_element);
}

Node *Node_new(void *element) {
    Node *node = (Node *)malloc(NODE_SIZE);
    node->element = element;
    node->next = NULL;
    node->previous = NULL;
    return node;
}

typedef struct _DoublyLinkedList {
    Node *head;
    Node *tail;
    int size;
} DoublyLinkedList;

const size_t LIST_SIZE = sizeof(DoublyLinkedList);

void DoublyLinkedList_set_head(DoublyLinkedList *list, Node *node) {
    if (list->head != NULL) {
        Node_set_previous(list->head, node);
    }
    list->head = node;
}

void DoublyLinkedList_set_tail(DoublyLinkedList *list, Node *node) {
    if (list->tail != NULL) {
        Node_set_next(list->tail, node);
    }
    list->tail = node;
}

bool DoublyLinkedList_is_empty(DoublyLinkedList *list) {
    return list->size < 1;
}

void DoublyLinkedList_append(DoublyLinkedList *list, Node *node) {
    DoublyLinkedList_set_tail(list, node);
    if (list->head == NULL) {
        DoublyLinkedList_set_head(list, node);
    }
    list->size++;
}

void DoublyLinkedList_prepend(DoublyLinkedList *list, Node *node) {
    DoublyLinkedList_set_head(list, node);
    if (list->tail == NULL) {
        DoublyLinkedList_set_tail(list, node);
    }
    list->size++;
}

void DoublyLinkedList_remove(DoublyLinkedList *list, Node *node, bool free_node) {
    Node *next = node->next;
    Node *previous = node->previous;
    if (next != NULL) {
        Node_set_previous(next, previous);
    } else if (previous != NULL) {
        Node_set_next(previous, NULL);
    }
    node->next = NULL;
    node->previous = NULL;
    list->size--;
    if (free_node) {
        free(node);
    }
}

DoublyLinkedList *DoublyLinkedList_new() {
    DoublyLinkedList *list = (DoublyLinkedList *)malloc(LIST_SIZE);
    list->head = NULL;
    list->tail = NULL;
    list->size = 0;
    return list;
}

void DoublyLinkedList_print(DoublyLinkedList *list) {
    Node *node = list->head;
    if (node == NULL) {
        return;
    }

    while (node != NULL) {
        Node_print(node);
        node = node->next;
    }
}

typedef struct _Stack {
    DoublyLinkedList *list;
} Stack;

const size_t STACK_SIZE = sizeof(Stack);

void *Stack_pop(Stack *stack) {
    Node *head = stack->list->head;
    if (head == NULL) {
        return NULL;
    }
    Node *next = head->next;
    DoublyLinkedList_remove(stack->list, head, false);
    DoublyLinkedList_set_head(stack->list, NULL);
    DoublyLinkedList_set_head(stack->list, next);
    void *element = head->element;
    free(head);
    return element;
}

void *Stack_top(Stack *stack) {
    Node *head = stack->list->head;
    return (head != NULL) ? head->element : NULL;
}

void Stack_push(Stack *stack, void *element) {
    Node *node = Node_new(element);
    DoublyLinkedList_prepend(stack->list, node);
}

Stack *Stack_new() {
    DoublyLinkedList *list = DoublyLinkedList_new();
    Stack *stack = (Stack *)malloc(STACK_SIZE);
    stack->list = list;
    return stack;
}

void cleanup(Stack *stack) {
    DoublyLinkedList *list = stack->list;
    Node *node = list->head;
    Node *next;
    while (node != NULL) {
        next = node->next;
        free(node);
        node = next;
    }
    free(list);
    free(stack);
}

int main() {
    Stack *stack = Stack_new();
    int e1 = 1;
    int e2 = 2;
    int e3 = 3;
    Stack_push(stack, &e1);
    Stack_push(stack, &e2);
    Stack_push(stack, &e3);
    printf("Size: %d\n", stack->list->size);
    int *val = (int *)Stack_pop(stack);
    printf("Pop value: %d\n", *val);
    printf("Size: %d\n", stack->list->size);
    printf("\n");
    DoublyLinkedList_print(stack->list);
    cleanup(stack);

    return 0;
}
